### ISSUE TITLE
streamingccl: range key splitting can produce incorrect SST

### ISSUE THREAD

**Issue by stevendanna** [cockroachdb/cockroach#112846]
When streaming receives a range key from the rangefeed-driven event stream, it adds them to an SST so they can be written at a specific timestamp. When calling AddSSTable for the built SST, the stream ingestion processor may need to split that SST into multiple SSTs if it is found to be at a range boundary.

There is a bug in the current splitting routine:

1) The streaming code inaccurately reports the end key of the SST by calling Next() on the actual end key.  This Next() is unnecessary because the EndKey of range keys are already exclusive.

2) When splitting an SST, the streaming code splits it at the split point returned by KV. We assume that the SST actually does contain values on the left- and right-hand side of the SST.

Given that behaviour, 

1) If we receive a range key for `[a, c)`, we will attempt to AddSSTable with bounds of `[a, c.Next())`.
2) If the destination side has a range split exactly at `c`, it will return a split point of c.
3) We will try to split the `[a, c)` range key at `c`, which will result in an empty right hand side.
4) Since our code assumes the rhs will be non-empty, we attempt to apply a bad SSTable and receive an error.

The stream can only advance past the error if the range boundaries happen to change at some point in the future.

Jira issue: CRDB-32637

**Comment by msbutler** [comment]
To clarify: 

Every SST created on the range key path will have an incorrect end key in its metadata, regardless of range key splitting, right?
https://github.com/cockroachdb/cockroach/blob/88a6fec61525b45c2fc4f6e0d11c0f54bdb33c68/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L923


If we remove the `Next()` call here, I don't think we'll need to worry about adding special logic to avoid an empty RHS SST: suppose we have a range key with the correct bounds of: `[a,c.Next())`, as in the range key covers all MVCC keys from `[a,c]`.
- If the range is split at `c`, then our current logic will create `[a,c)` and `[c,c.Next())` no problem.
- If the range is split at `c.Next()`, then the original range key doesn't need to be split. 

In other words, to fix this bug, all we need to do is remove the `.Next()` call [here](https://github.com/cockroachdb/cockroach/blob/88a6fec61525b45c2fc4f6e0d11c0f54bdb33c68/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go#L923), I think?

**Comment by adityamaru** [comment]
This bug does exist on release-23.1 too but we do not want to block the 23.1 release for this issue as the feature is not GA on that branch. I am dropping the branch label so that it stops showing up in the release blocker dashboards for 23.1 but we should backport the fix.

**Comment by stevendanna** [comment]
@msbutler Yes, we just need to review the Next(). And it is true that in the existing code, we push SSTs with slightly incorrect bounds. I'm unsure on the consequence of that. I mention splitting in the issue because when the bad metadata interacts with the splitting code that doesn't know how to handle it we end up with an SST that can't be applied because the bounds include nil keys.

For the fix, I am going to remove the Next() but also add some assertions so that we fail sooner if it looks like we've constructed something nonsensical.
